### PR TITLE
Harden key wrapping

### PR DESCRIPTION
**Pull request by ifranzki**
Cleanse areas that contain keys before freeing them.
Also don't use realloc to increase areas that contain keys.

This fixes issue https://github.com/opencryptoki/opencryptoki/issues/32

### CODE REVIEW

File: usr/lib/pkcs11/common/key_mgr.c [hunk 1]
```diff
@@ -1014,16 +1016,20 @@ CK_RV key_mgr_unwrap_key(STDLL_TokData_t *tokdata,
...
 
     return rc;
 
 error:
     if (key_obj)
         object_free(key_obj);
-    if (data)
+    if (data) {
         free(data);
+        OPENSSL_cleanse(data, data_len);
```
Comment (p-steuer) [inline usr/lib/pkcs11/common/key_mgr.c]: cleanse *after* free ?

Comment (ifranzki) [reply usr/lib/pkcs11/common/key_mgr.c]: Oups… :-)
Should be fixed now.

File: usr/lib/pkcs11/common/mech_aes.c [hunk 2]
```diff
@@ -3145,12 +3148,21 @@ CK_RV ckm_aes_wrap_format(STDLL_TokData_t *tokdata,
...
+             * Don't use realloc here, since the buffer contains a key and the
+             * old buffer needs to be cleansed before it is freed.
+             */
+            ptr = (CK_BYTE *)malloc(len2);
             if (!ptr) {
                 TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
                 return CKR_HOST_MEMORY;
             } else {
                 memset(ptr + len1, 0x0, (len2 - len1));
+                if (*data != NULL) {
```
Comment (p-steuer) [inline usr/lib/pkcs11/common/mech_aes.c]: can you remove the above "else" i.e. no "else" after "if {... return}"?

File: usr/lib/pkcs11/common/mech_des.c [hunk 3]
```diff
@@ -1459,12 +1462,21 @@ CK_RV ckm_des_wrap_format(STDLL_TokData_t *tokdata,
...
+             * Don't use realloc here, since the buffer contains a key and the
+             * old buffer needs to be cleansed before it is freed.
+             */
+            ptr = (CK_BYTE *)malloc(len2);
             if (!ptr) {
                 TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
                 return CKR_HOST_MEMORY;
             } else {
                 memset(ptr + len1, 0x0, (len2 - len1));
+                if (*data != NULL) {
```
Comment (p-steuer) [inline usr/lib/pkcs11/common/mech_des.c]: same here

File: usr/lib/pkcs11/common/mech_aes.c (hunk 2 above)
Comment (ifranzki) [reply usr/lib/pkcs11/common/mech_aes.c]: fixed

File: usr/lib/pkcs11/common/mech_des.c (hunk 3 above)
Comment (ifranzki) [reply usr/lib/pkcs11/common/mech_des.c]: fixed

**Pull request merged**